[flang][debug] Fix issues with local variables.

abidh opened the pull request:
This PR fixes 2 similar issues.
1. As reported in #97476, flang generated executable has inconsistent behavior regarding values of the local array variables.
2. Variable with save attribute would not show up in debugger.

The reason behind is same for both cases. If a local variable has storage which extends beyond function lifetime, the way to represent it in the debug info is through a global variable whose scope is limited to the function. This is what is used for static local variable in C. Previously local array worked in cases they were on stack. But will not show up if they had a global storage.

To fix this, if we can get a corresponding `GlobalOp` for a variable while processing `DeclareOp`, we treat it the variable as global with scope set appropriately. A new FIR test is added. A previous Integration test has been adjusted as to not expect local variables for local arrays.

With this fix in place, all the issues described in #97476 go away. It also fixes a lot of fails in GDB's fortran testsuite.

Fixes #97476.

abidh committed:
[flang][debug] Fix issues with local variables.

This PR fixes 2 similar issues.
1. As reported in #97476, flang generated executable has inconsistent
behavior regarding values of the local array variables.
2. Variable with save attribute would not show up in debugger.

The reason behind is same for both cases. If a local variable has
storage which extends beyond function lifetime, the way to represent
it in the debug info is through a global variable whose scope is limited
to the function. This is what is used for static local variable in C.
Previously local array worked in cases they were on stack. But will not
show up if they had a global storage.

To fix this, if we can get a corresponding GlobalOp for a variable while
processing DeclareOp, we treat it as a global variable.

jeanPerier commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -80,12 +80,19 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
     return;
 
+  // If this DeclareOp actually represents a global then treat it as such.
+  if (auto global = module.lookupSymbol<fir::GlobalOp>(declOp.getUniqName())) {
+    handleGlobalOp(global, fileAttr, scopeAttr);
```
module lookups are linear with the number of symbol inside the module.
 So this will introduce a "number of variables" * "number of types + number of global + number of function" pseudo quadratic behavior that is noticeable on real applications.

I advise sharing a const mlir::SymbolTable that would be initialized in in runOnOperation only once (ideally, it would be done in pass initialization, but I am not sure how it can be done with the initialize hook from Pass.h that only provide the mlir::Context).

abidh replied:
Thanks. That is very useful thing to know. I have updated the patch.

jeanPerier commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -139,6 +146,8 @@ mlir::LLVM::DIModuleAttr AddDebugInfoPass::getOrCreateModuleAttr(
 void AddDebugInfoPass::handleGlobalOp(fir::GlobalOp globalOp,
                                       mlir::LLVM::DIFileAttr fileAttr,
                                       mlir::LLVM::DIScopeAttr scope) {
+  if (mlir::isa<mlir::FusedLoc>(globalOp.getLoc()))
+    return;
```
Could you share that into a `debugInfoIsAlreadySet(globalOp.getLoc())` static helper so that there it is updated too when https://github.com/llvm/llvm-project/pull/95862 lands.

abidh replied:
Done.

jeanPerier reviewed (commented):
thanks, small non functional comments inline.

abidh committed:
Handle review comments.

1. Replace module.lookupSymbol with symbolTable->lookup for performance.

2. Move check that debug info is already processed in a separate function.

tblah reviewed (approved):
LGTM, thanks.

jeanPerier reviewed (approved):
Thanks!
